File: new/Basal.cpp
#include <iostream>
#include <string>
#include <cmath>
#include <vector>
#include <variant>
#include <algorithm>

using namespace std;
using numbas_model = vector<variant<double, char>>;

class Basal
{
public:
    int echo = 2, ripple = 0, polar = 2, discretion = 0, accuracy = 30;
    double velocity = 1;

    double num_value = 0;
    double base_value = 10;
    numbas_model number = {};
    string display = "<10T";

    Basal() {}
    Basal(double bass, double vel, int pol, int dis, int eco, int rip_val, int acc)
    {
        base_value = bass;
        velocity = vel;
        polar = pol;
        discretion = dis;
        echo = eco;
        ripple = rip_val;
        accuracy = acc;
    }
    Basal(numbas_model num, double bass, double vel, int pol, int dis, int eco, int rip_val, int acc)
    {
        number = num;
        base_value = bass;
        velocity = vel;
        polar = pol;
        discretion = dis;
        echo = eco;
        ripple = rip_val;
        accuracy = acc;

        Deconvert(num);
    }

    void Convert(int noom)
    {
        number = notobasten(noom, base_value, velocity, accuracy);

        if (polar != 2)
            number = depolaris(number, base_value, polar, velocity, accuracy);

        if (echo != 2)
            number = echor(ripple, echo, number);
    }
    void Deconvert(numbas_model number)
    {
        if (echo != 2)
            number = unechor(ripple, number, echo);

        if (polar != 2)
            num_value = polariser(number, base_value, polar, velocity);

        else
            num_value = tobasten(number, base_value, velocity);
    }

    double logbas(double x, double bass)
    {
        return log(x) / log(bass);
    }
    double expor(double num, double expo)
    {
        if (num < 0)
        {
            if (fmod(expo, 2) == 1)
                return (0 - pow(abs(num), expo));

            else
                return (pow(abs(num), expo));
        }
        else
            return (pow(num, expo));
    }

    int tobe(numbas_model num, char item)
    {
        for (int i = 0; i < num.size(); i++)
        {
            if (get<char>(num[i]) == item) { return i; }
            else { continue; }
        }

        return -1;
    }
    numbas_model substr(numbas_model sauce, int pos, int sizzly)
    {
        numbas_model outout = {};

        for (int i = 0; i < sizzly; i++)
        {
            outout.emplace_back(sauce.at(i + pos));
        }

        return outout;
    }

    double tobasten(numbas_model num, double bass, double velocity)
    {
        double outout = 0;

        int po = 0;
        if (get<char>(num[num.size() - 1]) == '-')
        {
            po = 1;
        }
        int dec = tobe(num, '.');

        if (dec != -1)
        {
            numbas_model noom_left = substr(num, 0, dec);
            numbas_model noom_right = substr(num, dec, num.size() - dec);

            for (int z = 0; z < noom_left.size(); z++)
            {
                outout += (pow(bass, (z * velocity)) * get<double>(noom_left[z]));
            }
            for (int z = 0; z < noom_right.size(); z++)
            {
                outout += (pow(bass, (0 - ((z + 1) * velocity))) * get<double>(noom_right[z]));
            }
        }
        else {
            for (int z = 0; z < num.size(); z++)
            {
                outout += pow(bass, (z * velocity)) * get<double>(num[z]);
            }
        }

        return outout;
    }
    numbas_model notobasten(double num, double bass, double velocity, int accuracy)
    {
        int statis = 0;

        if (bass < 1)
        {
            bass = 1 / bass; statis = 1;
        }

        int ceil = (1 + floor(logbas(num, pow(bass, velocity))));
        numbas_model outout = {};

        for (int z = 0; z < ceil; z++)
        {
            double index = pow(bass, ((ceil - (z + 1)) * velocity));

            int t = floor(num / index);

            if (t >= 1)
            {
                num -= (t * index);
            }

            outout.emplace_back(t);
        }

        if (num == 0)
        {
            if (statis == 0)
                return outout;
            else
                outout.emplace_back(0); outout.emplace_back('.'); reverse(outout.begin(), outout.end());  return outout;
        }
        outout.emplace_back('.');

        for (int z = 0; z < (accuracy - ceil); z++)
        {
            if (num == 0)
                return outout;

            double index = pow(bass, ((0 - (z + 1)) * velocity));

            int t = floor(num / index);

            if (t >= 1)
                num -= t * index;

            outout.emplace_back(t);
        }

        if (statis == 0)
            return outout;
        else
            reverse(outout.begin(), outout.end()); return outout;
    }

    double polariser(numbas_model num, double bass, int mod, double velocity)
    {
        double outout = 0;
        int l = num.size();
        int q = l - 1;

        if (tobe(num, '.') != -1)
            q -= tobe(num, '.');

        for (int z = 0; z < l; z++)
        {
            if (get<char>(num[z]) == '.')
                continue;

            q--;

            if (z % 2 == mod)
                outout += get<double>(num[z]) * pow(bass, (q * velocity));
            else
                outout -= get<double>(num[z]) * pow(bass, (q * velocity));
        }
        return outout;
    }
    numbas_model depolaris(numbas_model num, double bass, int mod, double velocity, int accuracy)
    {
        double albs = 0;
        int z = 0;
        reverse(num.begin(), num.end());


        if (tobe(num, '.' != -1))
            z -= tobe(num, '.');

        for (int z_2 = 0; z_2 < num.size(); z_2++)
        {
            if (get<char>(num[z]) == '.')
                continue;

            z += 1;

            if (get<double>(num[z]) != 0)
                albs += pow(bass, ((z + 1) * velocity)) + ((bass - (2 * get<double>(num[z]))) * pow(bass, (z * velocity)));
        }

        reverse(num.begin(), num.end());

        return notobasten(tobasten(num, bass, velocity) + albs, bass, velocity, accuracy);
    }

    numbas_model unechor(int ripple, numbas_model num, int mode)
    {
        int r = abs(ripple);
        int nlen = num.size();

        int mid;

        if (tobe(num, 'জ') != -1)
            mid = tobe(num, 'জ');

        else if (tobe(num, '&') != -1)
            mid = tobe(num, '&');

        else
            return {'w', 'h', 'y'};

        numbas_model left_arr = substr(num, 0, mid); reverse(left_arr.begin(), left_arr.end());
        numbas_model right_arr = substr(num, nlen - mid, num.size() - (nlen + mid));
        int fire = 73;
        int rx = right_arr.size(), rount = right_arr.size();
        int lx = left_arr.size(), lount = left_arr.size();
        numbas_model outout = {};

        if (ripple < 0)
        {
            reverse(right_arr.begin(), right_arr.end());
            reverse(left_arr.begin(), left_arr.end());
        }
        if (mode == 1)
            fire = 780;

        for (int z = 0; z < nlen; z++)
        {
            if (fire == 73)
            {
                for (int z_2 = 0; z_2 < r; z_2++)
                {
                    if (rx == 0)
                        break;

                    outout.emplace_back(right_arr[rount - rx]);
                    rx--;
                }
                fire = 780;
            }

            else if (fire == 780)
            {
                for (int z_2 = 0; z_2 < r; z_2++)
                {
                    if (lx == 0)
                        break;

                    outout.emplace_back(left_arr[lount - lx]);
                    lx--;
                }
                fire = 73;
            }
        }
        reverse(outout.begin(), outout.end());
        return outout;
    }
    numbas_model unechor(int ripple, numbas_model num, int mode, int mid)
    {
        int r = abs(ripple);
        int nlen = num.size();

        numbas_model left_arr = substr(num, 0, mid); reverse(left_arr.begin(), left_arr.end());
        numbas_model right_arr = substr(num, nlen - mid, num.size() - (nlen + mid));
        int fire = 73;
        int rx = right_arr.size(), rount = right_arr.size();
        int lx = left_arr.size(), lount = left_arr.size();
        numbas_model outout = {};

        if (ripple < 0)
        {
            reverse(right_arr.begin(), right_arr.end());
            reverse(left_arr.begin(), left_arr.end());
        }
        if (mode == 1)
            fire = 780;

        for (int z = 0; z < nlen; z++)
        {
            if (fire == 73)
            {
                for (int z_2 = 0; z_2 < r; z_2++)
                {
                    if (rx == 0)
                        break;

                    outout.emplace_back(right_arr[rount - rx]);
                    rx--;
                }
                fire = 780;
            }

            else if (fire == 780)
            {
                for (int z_2 = 0; z_2 < r; z_2++)
                {
                    if (lx == 0)
                        break;

                    outout.emplace_back(left_arr[lount - lx]);
                    lx--;
                }
                fire = 73;
            }
        }
        reverse(outout.begin(), outout.end());
        return outout;
    }
    numbas_model echor(int ripple, int mode, numbas_model num)
    {
        int r = abs(ripple);
        numbas_model left_arr = {};
        numbas_model right_arr = {};

        int x = num.size(), count = num.size();
        int fire = 73;

        if (mode == 1)
            fire = 780;

        while (x >= 0)
        {
            if (fire == 780)
                fire = 73;
            else if (fire == 73)
                fire = 780;

            for (int z = 0; z < r; z++)
            {
                if (fire == 73)
                {
                    if (x != 0)
                        left_arr.emplace_back(num.at(count - x));
                    else
                        break;
                }
                else if (fire == 780)
                {
                    if (x != 0)
                        right_arr.emplace_back(num.at(count - x));
                    else
                        break;
                }
                x--;
            }
        }

        if (ripple < 0)
            reverse(right_arr.begin(), right_arr.end());
        else
            reverse(left_arr.begin(), left_arr.end());

        left_arr.emplace_back('জ');
        left_arr.reserve(left_arr.size() + right_arr.size());
        left_arr.insert(left_arr.end(), right_arr.begin(), right_arr.end());
        return left_arr;
    }
};

int main()
{
    Basal bass;

    cout << bass.base_value;
}

File: new/Basal.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices.Swift;
using System.Text;

namespace Calculator
{
    internal class Basal
    {
        #region REQUIRED
        public static double expor(double num, double expo)
        {
            if (num < 0)
            {
                if (expo % 2 == 1)
                {
                    return (0 - Math.Pow(Math.Abs(num), expo));
                }
                else
                {
                    return (Math.Pow(Math.Abs(num), expo));
                }
            }
            else
            {
                return (Math.Pow(num, expo));
            }
        }

        #endregion

        #region CONVERT TO AND FROM DENARY

        public decimal tobasten(Array num, double bass, double velocity)
        {
            int po = 0;
            if (num[0] == '-')
            {
                po = 1;
                num = num.Substring(1);
            }
            decimal outout = 0;
            Array.Reverse(num);

            if (num.Contains('.'))
            {
                int pos = Array.IndexOf(num, '.');
                for (int z = 0; z < num.Length; z++)
                {
                    if (num[z] == '.') { pos++; continue; }
                    outout += (decimal)expor(bass, (z - pos) * velocity) * num[z];
                }
            }
            else
            {
                for (int z = 0; z < num.Length; z++)
                {
                    outout += (decimal)expor(bass, z * velocity) * num[z];
                }
            }

            if (po == 0) return outout;
            else return 0 - outout;
        }
        public string notobasten(decimal num, double bass, double velocity, int accuracy)
        {
            int statis = 0;
            if (bass < 1)
            {
                statis = 1;
                bass = 1 / bass;
            }

            int ceil = (int)(1 + Math.Floor(Math.Log(Math.Abs((double)num), bass * velocity)));
            object[] outout = {};

            for (int z = 0; z < ceil; z++)
            {
                int q = ceil - (z + 1);
                decimal index = (decimal)expor(bass, (q * velocity));

                int t = (int)(Math.Floor(num / index));

                if (t >= 1)
                {
                    num -= t * index;
                }
                outout.Append(t);
            }

            if (num == 0)
            {
                if (statis == 0) return outout;
                else
                {
                    Array.Reverse(outout);
                    return outout;
                }
            }
            else
            {
                outout.Append('.');

                for (int z = 0; z < (accuracy - ceil); z++)
                {
                    if (num == 0)
                    {
                        return outout;
                    }
                    decimal index = 1M;
                    try
                    {
                        index = (decimal)expor(bass, (0 - (z + 1)) * velocity);
                    }
                    catch
                    {
                        break;
                    }

                    int t = (int)(Math.Floor(num / index));

                    if (t >= 1)
                    {
                        num -= t * index;
                    }
                    outout.Append(t);
                }
                if (statis == 0) return outout;
                else { Array.Reverse(outout); return outout;}
            }
        }

        #endregion

        #region POLARITY

        public decimal polaris(Array num, double bass, double velocity, int mod)
        {
            decimal outout = 0m;
            int l = num.Length;
            int q = l;

            for (int z = 0; z < l; z++)
            {
                if (num[z] == '.') continue;

                q--;
                if (z % 2 == mod)
                {
                    outout += (decimal)(num[z] * expor(bass, (q * velocity)));
                }
                else
                {
                    outout -= (decimal)(num[z] * expor(bass, (q * velocity)));
                }
            }

            return outout;
        }
        public string depolaris(Array num, double bass, double velocity, int mod, int accuracy)
        {
            decimal abls = 0;

            string x = revstr(num);

            int z = 0;
            if (x.Contains('.'))
                z = x.IndexOf('.') - x.Length;

            for (int z_2 = 0; z_2 < x.Length; z_2++)
            {
                if (x[z_2] == '.') continue;

                z += 1;
                if (z % 2 == mod)
                {
                    if (x[z_2] != '0')
                    {
                        abls += (decimal)expor(bass, (z + 1) * velocity);
                        abls += (decimal)(bass - (2 * Cc(x[z_2])) * expor(bass, z * velocity));
                    }
                }
            }

            return notobasten(tobasten(num, bass, velocity) + abls, bass, velocity, accuracy);
        }
        #endregion

        #region ECHO

        public Array unechor(int rip, Array num, int mode)
        {
            object[] outout = {};
            int r = Math.Abs(rip);
            int nlen = num.Length - 1; int mid;

            try { mid = num.IndexOf('জ'); }
            catch { return "Error"; }

            object[] left = new ArraySegment(num, 0, mid);
            object[] right = new ArraySegment(num, mid + 1, num.Length - (mid + 1));
            string fire = "I";
            int rx = right.Length; int rount = rx;
            int lx = left.Length;int lount = lx;

            if (rip > 0)
                Array.Reverse(left);
            else
                Array.Reverse(right);

            if (mode == 1) fire = "780";

            for (int z = 0; z < nlen; z++)
            {
                if (fire == "I")
                {
                    for (int z_2 = 0; z_2 < r; z_2++)
                    {
                        if (rx == 0) break;

                        outout.Append(right[rount - rx]);
                        rx--;
                    }
                    fire = "780";
                }
                else if (fire == "780")
                {
                    for (int z_2 = 0; z_2 < r; z_2++)
                    {
                        if (lx == 0) break;

                        outout.Append(right[lount - lx]);
                        lx--;
                    }
                    fire = "I";
                }
            }

            return outout;
        }
        public Array unechor(int rip, Array num, int mode, int mid)
        {
            object[] outout = {};
            int r = Math.Abs(rip);
            int nlen = num.Length - 1;

            object[] left = new ArraySegment(num, 0, mid);
            object[] right = new ArraySegment(num, mid + 1, num.Length - (mid + 1));
            string fire = "I";
            int rx = right.Length; int rount = rx;
            int lx = left.Length; int lount = lx;

            if (rip > 0)
                Array.Reverse(left);
            else
                Array.Reverse(right);

            if (mode == 1) fire = "780";

            for (int z = 0; z < nlen; z++)
            {
                if (fire == "I")
                {
                    for (int z_2 = 0; z_2 < r; z_2++)
                    {
                        if (rx == 0) { break; }

                        outout.Append(right[rount - rx]);
                        rx--;
                    }
                    fire = "780";
                }
                else if (fire == "780")
                {
                    for (int z_2 = 0; z_2 < r; z_2++)
                    {
                        if (lx == 0) { break; }

                        outout.Append(right[lount - lx]);
                        lx--;
                    }
                    fire = "I";
                }
            }

            return outout;
        }
        public Array echor(string num, int mode, int rip, int uni)
        {
            int ripl = Math.Abs(rip);
            object[] left_arr = {};
            object[] right_arr = {};
            int x = num.Length;
            int count = x;
            char fire = 'I';
            char[]? placehold = ['জ'];

            if (uni == 1)
                placehold[0] = '&';

            while (x >= 0)
            {
                if (fire == "780") { fire = "I"; }
                else if (fire == "I") { fire = "780"; }
                else { return "How?"; }

                for (int z = 0; z < ripl; z++)
                {
                    if (fire == "I")
                    {
                        if (x == 0) break;
                        else
                        {
                            left_arr.Append(num[count - x]);
                        }
                    }

                    else if (fire == "780")
                    {
                        if (x == 0) break;
                        else
                        {
                            right_arr.Append(num[count - x]);
                        }
                    }
                    x--;
                }
                if (x == 0) { break; }
            }
            
            Array.Reverse(left_arr);
            object[] outout = {};

            if (mode == 0)
            {
                if (rip < 0)
                {
                    Array.Reverse(left_arr);
                    Array.Reverse(right_arr);
                }
                if (uni == 0) left_arr.Append('জ');
                else if (uni == 1) left_arr.Append('&');

                outout = left_arr.Concat(right_arr);
            }
            else
            {
                if (rip < 0)
                {
                    Array.Reverse(left_arr);
                    Array.Reverse(right_arr);
                }
                if (uni == 0) right_arr.Append('জ');
                else if (uni == 1) right_arr.Append('&');

                outout = right_arr.Concat(left_arr);
            }

            return outout;
        }

        #endregion

        #region DISPLAY

        public string? disp()
        {
            string outout = $"<{base_value}";

            if (polar != 2)
            {
                if (polar == 0)
                {
                    outout += "ঋ";
                }
                else if (polar == 1)
                {
                    outout += "দ";
                }
            }
            if (discretion != 0)
            {
                outout += "গ";
            }
            if (echo != 2)
            {
                if (echo == 0)
                {
                    outout += $"+ল{ripple}";
                }
                else if (echo == 1)
                {
                    outout += $"-ল{ripple}";
                }
            }
            if (velocity != 1)
            {
                outout += $"স{velocity}";
            }

            display = outout + "T";

            return display;
        }

        #endregion

        #region CALCULUS
 
        public decimal disintegrate(Array num, Basal inbas)
        {
            if (inbas.echo != 2) { num = inbas.unechor(inbas.ripple, num, inbas.echo); }

            if (inbas.polar != 2)
            {
                decimal outout = 0M ; int l = num.Length ; int q = l - 2;

                for (int z = 0; z < l; z++)
                {
                    if (num[z] == '.') { continue ;}
                    q--;
                    if (z % 2 == inbas.polar)
                        outout += num[z] * expor(inbas.base_value, (q - 1) * inbas.velocity) * q;
                    else
                        outout += num[z] * expor(inbas.base_value, (q - 1) * inbas.velocity) * q;
                }
                return outout;
            }
            else
            {
                po = 0;
                if (num[0] == '-') { num = new ArraySegment(num, 1, num.Length - 1) ; po = 1; }
                decimal outout = 0M; Array.Reverse(num); 

                if (num.Contains('.'))
                {
                    int pos = Array.IndexOf(num, '.');
                    for (int z = 0; z < num.Length; z++)
                    {
                        if (num[z] == '.') { pos++; continue; }
                        outout += (decimal)expor(bass, (z - pos - 1) * velocity) * num[z] * z;
                    }
                }
                else
                {
                    for (int z = 0; z < num.Length; z++)
                    {
                        outout += (decimal)expor(bass, (z - 1) * velocity) * num[z] * z;
                    }
                }

                if (po == 0)
                    return outout;
                else
                    return 0 - outout;
            }
        }

        #endregion

        #region ATTRIBUTES
        public int type = 1;
        public int polar = 2;
        public int discretion = 0;
        public int echo = 2;
        public int ripple = 0;
        public double velocity = 1;
        public int po = 0;

        public double base_value = 10;

        public string? display = "<10T";

        public object[] number = {'0'};
        public decimal value = 0;
        #endregion

        public Basal() { }

        public Basal(
            object[] num,
            double bass,
            int pol,
            int dis,
            int eco,
            int rip,
            double vel
            )
        {
            number = num;
            base_value = bass;
            polar = pol;
            discretion = dis;
            echo = eco;
            ripple = rip;
            velocity = vel;

            if (echo != 2)
            {
                num = unechor(rip, num, echo);
            }
            if (pol != 2)
            {
                value = polaris(num, bass, vel, pol);
            }
            else
            {
                value = tobasten(num, bass, vel);
            }
        }

        public Basal(
            double bass,
            int pol,
            int dis,
            int eco,
            int rip,
            double vel
            )
        {
            base_value = bass;
            polar = pol;
            discretion = dis;
            echo = eco;
            ripple = rip;
            velocity = vel;
        }

        public Basal(
            decimal num,
            double bass,
            int pol,
            int dis,
            int eco,
            int rip,
            double vel
            )
        {
            value = num;
            base_value = bass;
            polar = pol;
            discretion = dis;
            echo = eco;
            ripple = rip;
            velocity = vel;

            number = notobasten(num, bass, vel, 69);

            if (polar != 2)
            {
                number = depolaris(number, bass, vel, pol, 69);
            }

            if (echo != 2)
            {
                number = echor(number, echo, rip, 0);
            }
        }
    }
}